File: Estruturas de Dados e Complexidade de Algoritmos/Algoritmos33/Alg33.java
/*
♦ Algoritmo 33: Estrutura de Dados Não Linear Dinâmica (não Estática) de Árvore Binária: Árvore de Busca Binária (Binary Search Tree - BST).
*/


class No<T> {
    T valor;
    No<T> esquerda, direita;

    public No(T valor) {
        this.valor = valor;
        this.esquerda = this.direita = null;
    }
}

public class Alg33<T extends Comparable<T>> {
    private No<T> raiz;

    public Alg33() {
        this.raiz = null;
    }

    public void inserir(T valor) {
        this.raiz = inserirRecursivo(this.raiz, valor);
    }

    private No<T> inserirRecursivo(No<T> no, T valor) {
        if (no == null) {
            return new No<T>(valor);
        }

        if (valor.compareTo(no.valor) < 0) {
            no.esquerda = inserirRecursivo(no.esquerda, valor);
        } else if (valor.compareTo(no.valor) > 0) {
            no.direita = inserirRecursivo(no.direita, valor);
        }

        return no;
    }

    public boolean buscar(T valor) {
        return buscarRecursivo(this.raiz, valor);
    }

    private boolean buscarRecursivo(No<T> no, T valor) {
        if (no == null) {
            return false;
        }

        if (valor.compareTo(no.valor) == 0) {
            return true;
        }

        return valor.compareTo(no.valor) < 0 ? buscarRecursivo(no.esquerda, valor) : buscarRecursivo(no.direita, valor);
    }

    public void imprimirEmOrdem() {
        imprimirEmOrdemRecursivo(this.raiz);
    }

    private void imprimirEmOrdemRecursivo(No<T> no) {
        if (no != null) {
            imprimirEmOrdemRecursivo(no.esquerda);
            System.out.print(no.valor + " ");
            imprimirEmOrdemRecursivo(no.direita);
        }
    }

    public static void main(String[] args) {
        Alg33<Integer> arvore = new Alg33<>();
        arvore.inserir(10);
        arvore.inserir(5);
        arvore.inserir(20);
        arvore.inserir(3);
        arvore.inserir(8);
        arvore.inserir(15);

        System.out.println("Em Ordem:");
        arvore.imprimirEmOrdem();

        System.out.println("\nBuscar 8: " + arvore.buscar(8));
        System.out.println("Buscar 12: " + arvore.buscar(12));
    }
}

File: Estruturas de Dados e Complexidade de Algoritmos/Algoritmos33/Alg33.js
/*
♦ Algoritmo 33: Estrutura de Dados Não Linear Dinâmica (não Estática) de Árvore Binária: Árvore de Busca Binária (Binary Search Tree - BST).
*/


class No {
    constructor(valor) {
        this.valor = valor;
        this.esquerda = null;
        this.direita = null;
    }
}

class ArvoreBinaria {
    constructor() {
        this.raiz = null;
    }

    inserir(valor) {
        this.raiz = this.inserirRecursivo(this.raiz, valor);
    }

    inserirRecursivo(no, valor) {
        if (no === null) {
            return new No(valor);
        }

        if (valor < no.valor) {
            no.esquerda = this.inserirRecursivo(no.esquerda, valor);
        } else if (valor > no.valor) {
            no.direita = this.inserirRecursivo(no.direita, valor);
        }

        return no;
    }

    buscar(valor) {
        return this.buscarRecursivo(this.raiz, valor);
    }

    buscarRecursivo(no, valor) {
        if (no === null) {
            return false;
        }

        if (valor === no.valor) {
            return true;
        }

        return valor < no.valor ? this.buscarRecursivo(no.esquerda, valor) : this.buscarRecursivo(no.direita, valor);
    }

    imprimirEmOrdem() {
        this.imprimirEmOrdemRecursivo(this.raiz);
    }

    imprimirEmOrdemRecursivo(no) {
        if (no !== null) {
            this.imprimirEmOrdemRecursivo(no.esquerda);
            console.log(no.valor);
            this.imprimirEmOrdemRecursivo(no.direita);
        }
    }
}

const arvore = new ArvoreBinaria();
arvore.inserir(10);
arvore.inserir(5);
arvore.inserir(20);
arvore.inserir(3);
arvore.inserir(8);
arvore.inserir(15);

console.log("Em Ordem:");
arvore.imprimirEmOrdem();

console.log("\nBuscar 8: " + arvore.buscar(8));
console.log("Buscar 12: " + arvore.buscar(12));
